Comments DAV implementation

blizzz opened the pull request:
If you want to follow what is happening here.

Resolves #20264
- [x] If the user attempts to comment a file he has no access to, a 501 is thrown, should be 403.
## API spec

from https://github.com/owncloud/core/issues/20264#issuecomment-161742813

URL root is always remote.php/dav/comments/

| State | Function | Action | URL | Remark |
| --- | --- | :-: | :-: | --- |
| :checkered_flag: | create comment | POST | …/${entityType}/${entityId}/ | in future maybe also this url with several IDs appended (== replies) |
| :checkered_flag: | update comment | PROPPATCH | …/${entityType}/${entityId}/ ${commentId} |  |
| :checkered_flag: | delete comment | DELETE | …/${entityType}/${entityId}/ ${commentId} |  |
| :checkered_flag: | list comments | PROPFIND | …/${entityType}/${entityId} | The Comments Manager returns back the whole comments. ~~Filter restrictions (limit, offset, datetime) may be provided per request header.~~ Not possible with PROPFIND |
| :checkered_flag: | list comments | REPORT | …/${entityType}/${entityId} | The Comments Manager returns back the whole comments. Filter restrictions (limit, offset, datetime) may be provided per request header. |

blizzz committed:
Comments DAV implementation

PVince81 commented on apps/dav/lib/comments/commentsplugin.php:
```diff
@@ -0,0 +1,164 @@
+			$comment = $this->createComment($node->getName(), $node->getId(), $data, $request->getHeader('Content-Type'));
+			$url = $request->getUrl() . '/' . urlencode($comment->getId());
+
+			$response->setHeader('Location', $url);
```
Make this "Content-Location" (I was wrong originally and changed it in my pending tags PR)

blizzz replied:
thanks for the headsup :)

PVince81 commented on apps/dav/lib/comments/entitytypecollection.php:
```diff
@@ -0,0 +1,87 @@
+use Sabre\DAV\Exception\MethodNotAllowed;
+use Sabre\DAV\Exception\NotFound;
+
+class EntityTypeCollection extends Collection {
```
don't forget to add a comment, what is this collection and which child nodes does it contain :smile:

blizzz replied:
:ok_hand:

blizzz commented:
Another fix and creation does work :guitar: 

``` 
curl -i -u $USER:$PWD -X POST  -d '{"actorType":"users","verb":"comment","message":"first"}' -H "Content-Type: application/json" http://my.own.cloud/remote.php/dav/comments/files/$FILEID
```

blizzz commented:
REPORT request:

```curl -u user:pwd -i --data-binary "@report.xml" -X REPORT -H "Content-Type: text/xml" http://zara.owncloud.bzoc/master/remote.php/dav/comments/files/2156```

report.xml

```xml
<?xml version="1.0" encoding="utf-8" ?>
<D:report xmlns:D="DAV:" xmlns:oc="http://owncloud.org/ns" >
  <oc:limit>5</oc:limit>
  <oc:offset>0</oc:offset>
  <oc:datetime>2016-01-18 22:10:30</oc:datetime>
</D:report>
```

Example Output (without headers):

```
<?xml version="1.0"?>
<d:multistatus xmlns:d="DAV:" xmlns:s="http://sabredav.org/ns" xmlns:cal="urn:ietf:params:xml:ns:caldav" xmlns:cs="http://calendarserver.org/ns/" xmlns:card="urn:ietf:params:xml:ns:carddav" xmlns:oc="http://owncloud.org/ns">
 <d:response>
  <oc:comment>
   <oc:id>3</oc:id>
   <oc:parentId>0</oc:parentId>
   <oc:topmostParentId>0</oc:topmostParentId>
   <oc:childrenCount>0</oc:childrenCount>
   <oc:message>first</oc:message>
   <oc:verb>comment</oc:verb>
   <oc:actorType>users</oc:actorType>
   <oc:actorId>master</oc:actorId>
   <oc:objectType>files</oc:objectType>
   <oc:objectId>2156</oc:objectId>
   <oc:creationDateTime>2016-01-18 22:01:16</oc:creationDateTime>
   <oc:latestChildDateTime>2016-01-20 14:01:24</oc:latestChildDateTime>
  </oc:comment>
 </d:response>
 <d:response>
  <oc:comment>
   <oc:id>2</oc:id>
   <oc:parentId>0</oc:parentId>
   <oc:topmostParentId>0</oc:topmostParentId>
   <oc:childrenCount>0</oc:childrenCount>
   <oc:message>first</oc:message>
   <oc:verb>comment</oc:verb>
   <oc:actorType>users</oc:actorType>
   <oc:actorId>master</oc:actorId>
   <oc:objectType>files</oc:objectType>
   <oc:objectId>2156</oc:objectId>
   <oc:creationDateTime>2016-01-18 22:01:12</oc:creationDateTime>
   <oc:latestChildDateTime>2016-01-20 14:01:24</oc:latestChildDateTime>
  </oc:comment>
 </d:response>
</d:multistatus>
```

blizzz commented:
As I understand there are still some requirements to be met for proper compatibility with the REPORT spec, but basically that's like it.

PVince81 commented:
Nice stuff. According to the spec, you need to have a "propstat" block inside the "response" tags: https://tools.ietf.org/html/rfc2518#section-12.9

CC @evert for further guidance if needed

evert commented:
Hey guys,

First: you should always reference [rfc4918](http://tools.ietf.org/html/rfc4918), no exceptions. It's a much newer (and also much clearer) document.

@PVince81 is right, the XML structure here doesn't make a lot of sense. I think there are potentially a two options:

1. If you like multistatus, really the most sensible structure for WebDAV is to create one node per comment. Then for every comment that your query returns, you can simply report 1 comment per propstat and return 1 or more properties for each comment. This is pretty much how CalDAV calendar-query, and almost all of the other WebDAV REPORT requests work. So you are in relatively good company. The drawback is that it might a bit more complex to implement as you do need to represent each comment as a single node. HOWEVER: the other positive implication is that you now have a 'sane' way to implement `DELETE`, `PROPPATCH`, etc.
2. Do away entirely with multistatus, and simply make `<oc:comment>` your top-level XML element.

PVince81 commented:
@evert thanks for your feedback.

I think every comment MUST be implemented as a node anyway because they would be accessed through other endpoints for creation/edition, etc

evert commented:
In that case, if you want things to work the most like the rest of WebDAV, it might be best to either:

1. Define a single `oc:comment` WebDAV property on all nodes,
2. Split the information up in individual WebDAV properties.

The first might be easier and faster, but the second is a bit more WebDAV-like. If you're going for a single `oc:comment` property, make sure you keep the existing `xmlSerialize` function. You can largely re-use it when you reimplement this as a WebDAV property.

The nice thing about WebDAV properties is that you can implement the property first and just test that using the browser plugin (is it enabled yet?? it really really should!). And then after that and you know it works perfectly, you can add the REPORT query which is really just an optimization....

blizzz commented:
Thanks, @evert, for the input!

Every comment is a node already :white_check_mark: 

> First: you should always reference rfc4918, no exceptions. It's a much newer (and also much clearer) document.

However specifically about REPORT method it does not contain much information. Or did I oversee anything? Nevertheless, thanks for the hint, already found useful info there.

> Define a single oc:comment WebDAV property on all nodes,

So basically what we have https://github.com/owncloud/core/pull/21664#issuecomment-173222186, only that it should be propperly embedded in a propstat structure, like https://tools.ietf.org/html/rfc4918#section-15.8.1

> The nice thing about WebDAV properties is that you can implement the property first and just test that using the browser plugin (is it enabled yet?? it really really should!). 

Temporarialy enabled it for me. This performes an PROPFIND, I assume, at least the output looks the same.

evert commented:
> Temporarialy enabled it for me. This performes an PROPFIND, I assume, at least the output looks the same.

Yes, so if the comment is properly implemented as a WebDAV property, it should also show up there. To do this, you'll want to implement `Sabre\DAV\IProperties` and not `Sabre\Xml\XmlSerializable`.

Specifically the `propFind` function is important.

evert commented:
Making sure that that's correct (both in the Browser as well as in PROPFIND) is what I think should be your first step. `REPORT` is then next. You are right that `REPORT` is not described in the original webdav specification, it comes from:

https://tools.ietf.org/html/rfc3253

But basically you should treat `REPORT` pretty much the same way as you treat `POST`. The difference is that `REPORT` must be idempotent and safe, whereas `POST` is neither.

blizzz commented:
Current State:

One thing I came across: when requiring all properties with propfind, sabre overrides it and sets it's own choice of properties in the PropFind constructor. Strange, but OK. 

Using the browser plugin and going to the comment node, the fictional  ``{http://sabredav.org/ns}idk`` property is requested. Hm. 

Using the browser plugin and going to the file collection (within the comments tree), I get to see my test comments and the links to them. That's good I guess.

Doing the propfind with curl and requesting our comment property and few properties from the comment itself , I'll get 

```xml
<?xml version="1.0"?>
<d:multistatus xmlns:d="DAV:" xmlns:s="http://sabredav.org/ns" xmlns:cal="urn:ietf:params:xml:ns:caldav" xmlns:cs="http://calendarserver.org/ns/" xmlns:card="urn:ietf:params:xml:ns:carddav" xmlns:oc="http://owncloud.org/ns">
 <d:response>
  <d:href>/master/remote.php/dav/comments/files/2156/</d:href>
  <d:propstat>
   <d:prop>
    <oc:comment>
     <oc:comment>3</oc:comment>
     <oc:comment>1</oc:comment>
    </oc:comment>
   </d:prop>
   <d:status>HTTP/1.1 200 OK</d:status>
  </d:propstat>
  <d:propstat>
   <d:prop>
    <oc:message/>
    <oc:actorId/>
    <oc:actorType/>
   </d:prop>
   <d:status>HTTP/1.1 404 Not Found</d:status>
  </d:propstat>
 </d:response>
 <d:response>
  <d:href>/master/remote.php/dav/comments/files/2156/3</d:href>
  <d:propstat>
   <d:prop>
    <oc:message>first</oc:message>
    <oc:actorId>master</oc:actorId>
    <oc:actorType>users</oc:actorType>
   </d:prop>
   <d:status>HTTP/1.1 200 OK</d:status>
  </d:propstat>
  <d:propstat>
   <d:prop>
    <oc:comment/>
   </d:prop>
   <d:status>HTTP/1.1 404 Not Found</d:status>
  </d:propstat>
 </d:response>
 <d:response>
  <d:href>/master/remote.php/dav/comments/files/2156/1</d:href>
  <d:propstat>
   <d:prop>
    <oc:message>message</oc:message>
    <oc:actorId>master</oc:actorId>
    <oc:actorType>users</oc:actorType>
   </d:prop>
   <d:status>HTTP/1.1 200 OK</d:status>
  </d:propstat>
  <d:propstat>
   <d:prop>
    <oc:comment/>
   </d:prop>
   <d:status>HTTP/1.1 404 Not Found</d:status>
  </d:propstat>
 </d:response>
</d:multistatus>
```

What I don't understand yet, is the second propstat block to each resource, that returns a 404.

> You are right that REPORT is not described in the original webdav specification, it comes from:
> https://tools.ietf.org/html/rfc3253

Seen this, rich on details ;)

PVince81 commented:
What is the "oc:comment" property supposed to contain ?
I think we don't need it at all, because the propstat result itself is already representing the comment.

blizzz commented:
> What is the "oc:comment" property supposed to contain ?

The comment data. The CommentNode provides all.

evert commented on apps/dav/lib/comments/commentnode.php:
```diff
@@ -0,0 +1,196 @@
+use Sabre\Xml\Writer;
+use Sabre\Xml\XmlSerializable;
+
+class CommentNode implements \Sabre\DAV\INode, XmlSerializable,\Sabre\DAV\IProperties {
```
I would definitely suggest dropping XmlSerializable

blizzz replied:
Sorry, I still miss to see how I would return this on REPORT without that. Leaving that out I am greeted with ``The writer cannot serialize objects of class: OCA\DAV\Comments\CommentNode``. (Yes, I still need to correct output).

evert commented:
> One thing I came across: when requiring all properties with propfind, sabre overrides it and sets it's own choice of properties in the PropFind constructor. Strange, but OK. 

`{DAV:}allprop` is weird. It was originally intended to return all properties, but if you read later versions of the webdav spec you'll see that this is no longer true. The reason is that many clients started using this, but for servers this was actually way to costly to calculate. So now clients are forced to specify only the properties they actually want.

> Using the browser plugin and going to the comment node, the fictional {http://sabredav.org/ns}idk property is requested. Hm. 

Even though WebDAV doesn't really support allprop, the browser plugin uses a bit of a hack to force the server to return everything anyway. This is part of that hack ;)

My suggestion:

if it's not 'expensive' I would simply return *everything you have* from `getProperties()`. The server will make sure to discard the properties that are not actually requested. So this will make your implementation a bit easier, and it will also allow you to see everything in the browser plugin.

blizzz commented:
@evert Thanks for the clarification on allprop!

>> Using the browser plugin and going to the comment node, the fictional {http://sabredav.org/ns}idk property is requested. Hm. 

> Even though WebDAV doesn't really support allprop, the browser plugin uses a bit of a hack to force the server to return everything anyway. This is part of that hack ;)

> My suggestion:

> if it's not 'expensive' I would simply return *everything you have* from `getProperties()`. The server will make sure to discard the properties that are not actually requested. So this will make your implementation a bit easier, and it will also allow you to see everything in the browser plugin.

It's not expensive, we'll have everything at hand at this point of runtime. Reducing this saves some other code, so I will go for it. However, the browser plugin does not show me more information regardless of it.

blizzz commented:
Took the low hanging stuff inbetween. DELETE on a comment  worked already, PROPPATCH for updating the comment's message (and only that, other details are not allowed), is added.

blizzz commented:
reworked the REPORT stuff, meanwhile I figured out how to get rid off XmlSerializable :) Basically I was shamelessly checking how it is done in Sabre\DAV\Sync\Plugin::sendSyncCollectionResponse().

Now, my output for ``curl -u user:pwd -i --data-binary "@report.xml" -X REPORT -H "Content-Type: text/xml" http://zara.owncloud.bzoc/master/remote.php/dav/comments/files/2156`` is:

```xml

<?xml version="1.0"?>
<x1:multistatus xmlns:x1="DAV:">
 <x1:response xmlns:x1="DAV:">
  <x1:status xmlns:x1="DAV:">HTTP/1.1 200 OK</x1:status>
  <x1:href xmlns:x1="DAV:">/master/remote.php/dav/http://zara.owncloud.bzoc/master/remote.php/dav/comments/files/2156</x1:href>
  <x1:propstat xmlns:x1="DAV:">
   <x1:prop xmlns:x1="DAV:">
    <x2:comment xmlns:x2="http://owncloud.org/ns">
     <x1:href xmlns:x1="DAV:">http://zara.owncloud.bzoc/master/remote.php/dav/comments/files/2156/6</x1:href>
     <x2:id xmlns:x2="http://owncloud.org/ns">6</x2:id>
     <x2:parentId xmlns:x2="http://owncloud.org/ns">0</x2:parentId>
     <x2:topmostParentId xmlns:x2="http://owncloud.org/ns">0</x2:topmostParentId>
     <x2:childrenCount xmlns:x2="http://owncloud.org/ns">0</x2:childrenCount>
     <x2:message xmlns:x2="http://owncloud.org/ns">Das hier ist Fußball, das hier sind Dramen</x2:message>
     <x2:verb xmlns:x2="http://owncloud.org/ns">comment</x2:verb>
     <x2:actorType xmlns:x2="http://owncloud.org/ns">users</x2:actorType>
     <x2:actorId xmlns:x2="http://owncloud.org/ns">master</x2:actorId>
     <x2:objectType xmlns:x2="http://owncloud.org/ns">files</x2:objectType>
     <x2:objectId xmlns:x2="http://owncloud.org/ns">2156</x2:objectId>
     <x2:creationDateTime xmlns:x2="http://owncloud.org/ns">2016-01-21 18:01:48</x2:creationDateTime>
     <x2:latestChildDateTime xmlns:x2="http://owncloud.org/ns">2016-01-21 19:01:32</x2:latestChildDateTime>
    </x2:comment>
   </x1:prop>
   <x1:status xmlns:x1="DAV:">HTTP/1.1 200 OK</x1:status>
  </x1:propstat>
 </x1:response>
 <x1:response xmlns:x1="DAV:">
  <x1:status xmlns:x1="DAV:">HTTP/1.1 200 OK</x1:status>
  <x1:href xmlns:x1="DAV:">/master/remote.php/dav/http://zara.owncloud.bzoc/master/remote.php/dav/comments/files/2156</x1:href>
  <x1:propstat xmlns:x1="DAV:">
   <x1:prop xmlns:x1="DAV:">
    <x2:comment xmlns:x2="http://owncloud.org/ns">
     <x1:href xmlns:x1="DAV:">http://zara.owncloud.bzoc/master/remote.php/dav/comments/files/2156/5</x1:href>
     <x2:id xmlns:x2="http://owncloud.org/ns">5</x2:id>
     <x2:parentId xmlns:x2="http://owncloud.org/ns">0</x2:parentId>
     <x2:topmostParentId xmlns:x2="http://owncloud.org/ns">0</x2:topmostParentId>
     <x2:childrenCount xmlns:x2="http://owncloud.org/ns">0</x2:childrenCount>
     <x2:message xmlns:x2="http://owncloud.org/ns">I want to play a game</x2:message>
     <x2:verb xmlns:x2="http://owncloud.org/ns">comment</x2:verb>
     <x2:actorType xmlns:x2="http://owncloud.org/ns">users</x2:actorType>
     <x2:actorId xmlns:x2="http://owncloud.org/ns">master</x2:actorId>
     <x2:objectType xmlns:x2="http://owncloud.org/ns">files</x2:objectType>
     <x2:objectId xmlns:x2="http://owncloud.org/ns">2156</x2:objectId>
     <x2:creationDateTime xmlns:x2="http://owncloud.org/ns">2016-01-21 18:01:47</x2:creationDateTime>
     <x2:latestChildDateTime xmlns:x2="http://owncloud.org/ns"/>
    </x2:comment>
   </x1:prop>
   <x1:status xmlns:x1="DAV:">HTTP/1.1 200 OK</x1:status>
  </x1:propstat>
 </x1:response>
 <x1:response xmlns:x1="DAV:">
  <x1:status xmlns:x1="DAV:">HTTP/1.1 200 OK</x1:status>
  <x1:href xmlns:x1="DAV:">/master/remote.php/dav/http://zara.owncloud.bzoc/master/remote.php/dav/comments/files/2156</x1:href>
  <x1:propstat xmlns:x1="DAV:">
   <x1:prop xmlns:x1="DAV:">
    <x2:comment xmlns:x2="http://owncloud.org/ns">
     <x1:href xmlns:x1="DAV:">http://zara.owncloud.bzoc/master/remote.php/dav/comments/files/2156/4</x1:href>
     <x2:id xmlns:x2="http://owncloud.org/ns">4</x2:id>
     <x2:parentId xmlns:x2="http://owncloud.org/ns">0</x2:parentId>
     <x2:topmostParentId xmlns:x2="http://owncloud.org/ns">0</x2:topmostParentId>
     <x2:childrenCount xmlns:x2="http://owncloud.org/ns">0</x2:childrenCount>
     <x2:message xmlns:x2="http://owncloud.org/ns">let us talk about files</x2:message>
     <x2:verb xmlns:x2="http://owncloud.org/ns">comment</x2:verb>
     <x2:actorType xmlns:x2="http://owncloud.org/ns">users</x2:actorType>
     <x2:actorId xmlns:x2="http://owncloud.org/ns">master</x2:actorId>
     <x2:objectType xmlns:x2="http://owncloud.org/ns">files</x2:objectType>
     <x2:objectId xmlns:x2="http://owncloud.org/ns">2156</x2:objectId>
     <x2:creationDateTime xmlns:x2="http://owncloud.org/ns">2016-01-21 18:01:47</x2:creationDateTime>
     <x2:latestChildDateTime xmlns:x2="http://owncloud.org/ns"/>
    </x2:comment>
   </x1:prop>
   <x1:status xmlns:x1="DAV:">HTTP/1.1 200 OK</x1:status>
  </x1:propstat>
 </x1:response>
</x1:multistatus>
```

Which is now almost acceptable, I think. Only, my gut feeling would be happier with just one response block, however I am getting different errors when I try that and am to tired to get behind this.

PVince81 commented:
- [ ] Something wrong with the href: `/master/remote.php/dav/http://zara.owncloud.bzoc/master/remote.php/dav/comments/files/2156`
- [ ] might be missing namespace mapping (x1, x2 tags)

Also, since you are calling REPORT on the files element, I thought returning a list of propstat with the contents of "oc:comment" would be enough. But maybe the spec requires the format you provided, so should be fine for now.

PVince81 commented:
The response seems to imply that you have a "file-comment mapping node" that itself has a "comment" property containing the comment.

I'd have structured it differently and instead of having an embedded "comment" property, I'd simply have said that the comments are the contents of the "files-mapping" collection instead of being a property.

If we compare this to files and folders, your implementation would have "folder contents aka sub-files" as an attribute of the "file node", which doesn't feel right. It is better to have the sub-files being the contents of the collection instead of an attribute. Hope this clarifies it a bit :smile: 

@evert what do you think ?

PVince81 commented on apps/dav/lib/comments/collection.php:
```diff
@@ -0,0 +1,178 @@
+use Sabre\DAV\Exception\NotFound;
+use Sabre\DAV\ICollection;
+
+class Collection implements ICollection {
```
EntityTypeCollection ? better be explicit

blizzz replied:
Root collection

PVince81 commented on apps/dav/lib/comments/collection.php:
```diff
@@ -0,0 +1,178 @@
+	}
+
+	/**
+	 * initializes the connection. At this point of time, we need the logged in
```
s/connection/collection/
I sense you've been doing too much LDAP lately :wink:

PVince81 commented on apps/dav/lib/comments/commentnode.php:
```diff
@@ -0,0 +1,173 @@
+		$result = [];
+		foreach($properties as $property) {
+			$getter = $this->properties[$property];
+			if(method_exists($this->comment, $getter)) {
```
so lazy :wink:

blizzz replied:
the getter methods actually come from the comment  instance itself, the method_exist check is just a defensive approach.

PVince81 commented on apps/dav/lib/comments/commentsplugin.php:
```diff
@@ -0,0 +1,249 @@
+	 * @return void
+	 */
+	function initialize(Server $server) {
+		$server->xml->namespaceMap[self::NS_OWNCLOUD] = 'oc';
```
hmm, that should solve the weird namespace you have in your last response.
Maybe you need to add something similar in other places ?

blizzz replied:
thanks for the hint, i was looking for something like that, will try it :)

blizzz replied:
:heavy_check_mark:

PVince81 commented on apps/dav/lib/comments/commentsplugin.php:
```diff
@@ -0,0 +1,249 @@
+		if ($node instanceof EntityCollection) {
+			$data = $request->getBodyAsString();
+
+			$comment = $this->createComment($node->getName(), $node->getId(), $data, $request->getHeader('Content-Type'));
```
You need to split the "Content-Type", some browsers send additional info.
See how I had to fix it here: https://github.com/owncloud/core/blob/master/apps/dav/lib/systemtag/systemtagplugin.php#L150

blizzz replied:
done

PVince81 commented on apps/dav/lib/comments/commentsplugin.php:
```diff
@@ -0,0 +1,249 @@
+			$ns = '{' . self::NS_OWNCLOUD . '}';
+			$writer->startElement($ns . 'comment');
+
+			$writer->writeElement('{DAV:}href',  $this->server->httpRequest->getAbsoluteUrl() . '/' . $node->comment->getId());
```
Hmm, I'd expect SabreDAV to provide a PropStat class that you just populate and let itself output itself.
Let me check

PVince81 replied:
I see there is a `Response` class that takes an array of properties. You could even go higher and use the `MultiStatus` class. I think one of these should take your `$comment` object as input and automatically output all requested properties.

PVince81 replied:
@evert is that correct ? Or is there an even higher level class that can be used as entry point for the response ?

PVince81 replied:
Ok, I see you already used `Multistatus` here.
The only reason you're doing this big block is because you have sub-properties.

Then we need to resolve the discussion I proposed about "sub-properties" vs "collection contents".

PVince81 commented:
If you would do a PROPFIND on http://zara.owncloud.bzoc/master/remote.php/dav/comments/files/2156 you would get a list of ALL comments, http://zara.owncloud.bzoc/master/remote.php/dav/comments/files/2156/{1,2,3,4}
and each of the results would not have a sub-property oc:comment but be the comment itself

So basically shifting the contents of "oc:comment" one level higher and getting rid of the "oc:comment" property.
The goal is to have the result of REPORT looks like the PROPFIND result but filtered/paginated.

This could help immensely for parsing the result later in the JS code by reusing the PROPFIND-parsing code.

blizzz commented on apps/dav/lib/comments/commentnode.php:
```diff
@@ -0,0 +1,196 @@
+			$name = '{'.self::NS_OWNCLOUD.'}' . lcfirst(substr($getter, 3));
+			$this->properties[$name] = $getter;
+		}
+		file_put_contents('/tmp/export', print_r(array_keys($this->properties), true));
```
ah, this must go :D

blizzz commented:
I reworked REPORT logic again based on discussion with @PVince81.  I was now really able to get rid of most of the serialization magic, except for adding DateTime to the classMap. However, on a global level so it works with PROPFIND: https://github.com/owncloud/core/pull/21664/files#diff-1ae5e0a1a5d93d278bb2f6831eed5ca6R91

* [ ] I think this can/should be moved to a beforeMethod handller to avoid conflicts.

Response is now similar to PROPFIND, which also means i got read of oc:comment either, comment data is now directly in a response block.

PVince81 commented:
That's some mighty stuff :smile:

PVince81 commented:
@blizzz regarding the classMap I think it should be done on property level instead to avoid potential conflicts with other plugins.

- [ ] the input datetime `<oc:datetime>2016-01-18 22:10:30</oc:datetime>` should be changed to use the same standard date format used for the output. Maybe you can somehow reuse the (de)serializer ?

PVince81 commented:
My test output looks like this, it's a nice and clean PROPFIND multistatus response:

```xml
<?xml version="1.0"?>
<d:multistatus xmlns:d="DAV:" xmlns:s="http://sabredav.org/ns" xmlns:cal="urn:ietf:params:xml:ns:caldav" xmlns:cs="http://calendarserver.org/ns/" xmlns:card="urn:ietf:params:xml:ns:carddav" xmlns:oc="http://owncloud.org/ns">
 <d:response>
  <d:status>HTTP/1.1 200 OK</d:status>
  <d:href>/owncloud/remote.php/dav/comments/files/3/3</d:href>
  <d:propstat>
   <d:prop>
    <oc:id>3</oc:id>
    <oc:parentId>0</oc:parentId>
    <oc:topmostParentId>0</oc:topmostParentId>
    <oc:childrenCount>0</oc:childrenCount>
    <oc:message>☻</oc:message>
    <oc:verb>comment</oc:verb>
    <oc:actorType>users</oc:actorType>
    <oc:actorId>root</oc:actorId>
    <oc:creationDateTime>Fri, 22 Jan 2016 14:00:03 GMT</oc:creationDateTime>
    <oc:latestChildDateTime/>
    <oc:objectType>files</oc:objectType>
    <oc:objectId>3</oc:objectId>
   </d:prop>
   <d:status>HTTP/1.1 200 OK</d:status>
  </d:propstat>
 </d:response>
 <d:response>
  <d:status>HTTP/1.1 200 OK</d:status>
  <d:href>/owncloud/remote.php/dav/comments/files/3/2</d:href>
  <d:propstat>
   <d:prop>
    <oc:id>2</oc:id>
    <oc:parentId>0</oc:parentId>
    <oc:topmostParentId>0</oc:topmostParentId>
    <oc:childrenCount>0</oc:childrenCount>
    <oc:message>second</oc:message>
    <oc:verb>comment</oc:verb>
    <oc:actorType>users</oc:actorType>
    <oc:actorId>root</oc:actorId>
    <oc:creationDateTime>Fri, 22 Jan 2016 13:57:00 GMT</oc:creationDateTime>
    <oc:latestChildDateTime/>
    <oc:objectType>files</oc:objectType>
    <oc:objectId>3</oc:objectId>
   </d:prop>
   <d:status>HTTP/1.1 200 OK</d:status>
  </d:propstat>
 </d:response>
 <d:response>
  <d:status>HTTP/1.1 200 OK</d:status>
  <d:href>/owncloud/remote.php/dav/comments/files/3/1</d:href>
  <d:propstat>
   <d:prop>
    <oc:id>1</oc:id>
    <oc:parentId>0</oc:parentId>
    <oc:topmostParentId>0</oc:topmostParentId>
    <oc:childrenCount>0</oc:childrenCount>
    <oc:message>first</oc:message>
    <oc:verb>comment</oc:verb>
    <oc:actorType>users</oc:actorType>
    <oc:actorId>root</oc:actorId>
    <oc:creationDateTime>Fri, 22 Jan 2016 13:56:34 GMT</oc:creationDateTime>
    <oc:latestChildDateTime/>
    <oc:objectType>files</oc:objectType>
    <oc:objectId>3</oc:objectId>
   </d:prop>
   <d:status>HTTP/1.1 200 OK</d:status>
  </d:propstat>
 </d:response>
</d:multistatus>
```

PVince81 commented:
elementMap doesn't seem to work, possible Sabre bug or misunderstood API: https://github.com/fruux/sabre-dav/issues/767

blizzz commented:
Now, before settting classMap for DateTime the RequestURI is checked, this should avoid conflicts. Being consequent, doing the other initializations can also be cancelled if the request runs against something else. This would avoid unnecessary roundtrips.

blizzz commented:
To complete the last requirement in https://github.com/owncloud/core/issues/20264, CommentNode also returns the display name of the actor, if the actor is a user. Since  it is theoretically possible that at some point we or someone else supports comments on Users, I called this property ``{http://owncloud.org/ns}actorDisplayName`` instead of reusing ``{DAV:}displayname`` which would not allow to discriminate against displayname of the target object and the actor.

blizzz commented:
Open for review @PVince81 @DeepDiver1975 @nickvergessen

PVince81 commented on apps/dav/lib/comments/commentnode.php:
```diff
@@ -0,0 +1,211 @@
+	 */
+	function getLastModified() {
+		// we do not have a separate "mDateTime" field for updates currently.
+		return $this->comment->getCreationDateTime()->getTimestamp();
```
This is the value returned when querying the "{DAV:}getlastmodified" field.
Is that intended ? If not, then make it return null or another invalid value.

blizzz replied:
Honestly, this comes from INode and I just returned that value, and originally not even the right one but DateTime. Without effect. So I guess we can kill this indeed.

PVince81 commented on apps/dav/lib/comments/commentnode.php:
```diff
@@ -0,0 +1,211 @@
+	function propPatch(PropPatch $propPatch) {
+		// other properties than 'message' are read only
+		$propPatch->handle('{'.self::NS_OWNCLOUD.'}message', [$this, 'updateComment']);
+		$propPatch->commit();
```
Is a commit required here ? Just asking because I didn't have it in my code :smile:

blizzz replied:
Yes. The handle() stuff is just telling what should happens with which attributes, but commit() triggers the update routines.

evert replied:
You should actually not call commit from your code. Bad news =) it's called elsewhere.

blizzz replied:
Hm, okay. Verified it really works without :) PR following.

PVince81 commented on apps/dav/lib/comments/commentnode.php:
```diff
@@ -0,0 +1,211 @@
+		$methods = array_filter($methods, function($name){
+			return strpos($name, 'get') === 0;
+		});
+		foreach($methods as $getter) {
```
Maybe should restrict this to allowed property names ?

Else someone could trick this class to call an arbitrary get method.
For example "getName()" by requesting "oc:name" which isn't intended.

blizzz replied:
There's nothing restricted to read from IComment.

PVince81 commented on apps/dav/lib/comments/commentsplugin.php:
```diff
@@ -0,0 +1,243 @@
+		try {
+			$node = $this->server->tree->getNodeForPath($path);
+		} catch (NotFound $e) {
+			return null;
```
If the node doesn't exist you should let the exception be thrown.
Basically just remove the try/catch block.

blizzz replied:
Well, I took it from https://github.com/owncloud/core/blob/master/apps/dav/lib/systemtag/systemtagplugin.php#L103 

But you're right, we can simplify this.

PVince81 replied:
Maybe I was wrong back then... returning null usually gives a chance for other plugins to handle this node.
But if the node doesn't exist in the first place it might be better to throw directly.

PVince81 commented on apps/dav/lib/comments/commentsplugin.php:
```diff
@@ -0,0 +1,243 @@
+		}
+
+		if(!is_null($args['datetime'])) {
+			$args['datetime'] = new \DateTime($args['datetime']);
```
So, automatic parsing of the date through elementMap did not work in the end ? (remember, evert said it should work on input)

PVince81 replied:
I'm fine with keeping this if it is able to parse the standard Webdav date format, same format as getlastmodified

blizzz replied:
Just added this, works.

blizzz replied:
… but kills unit test. I'd introduce a new method and need to write more tests about it. Makes things more complex and breakable, really. Reverted.

PVince81 commented on apps/dav/lib/comments/commentsplugin.php:
```diff
@@ -0,0 +1,243 @@
+				'message' => 'setMessage',
+				'verb' => 'setVerb',
+			];
+			foreach($properties as $property => $setter) {
```
This loop might be overkill just to set two properties ? Setting them directly might be more readable.
(optional)

blizzz replied:
Saves also few lines.

PVince81 commented:
Great work! The code looks good with a few minor comments.

Let me test the endpoints.

PVince81 commented:
- [ ] BUG or misunderstanding: if file with id "9" has no comments attached, I'd expect that if I do a PROPFIND on http://root:admin@localhost/owncloud/remote.php/dav/comments/files/9 it would return 404, at least that's how I did with system tags.

PVince81 commented:
- [ ] BUG: POST to a non-existing file id "http://root:admin@localhost/owncloud/remote.php/dav/comments/files/100 " must return 404, not 403 ?

PVince81 commented:
- [x] TEST: creating messages with POST
- [ ] TEST: create message on non-existing files => see above
- [x] TEST: delete file collection => must be forbidden
- [x] TEST: list messages on specific file with PROPFIND
- [ ] TEST: list messages on non-existing file must fail with 404 :no_entry_sign: => it returns 403, must be 404
- [x] TEST: list messages with REPORT
- [ ] TEST: list messages with REPORT on non-existing file, must fail with 404 :no_entry_sign: => it returns 403, must be 404
- [x] TEST: REPORT on existing file that has no comments => empty
- [x] TEST: delete existing message
- [x] TEST: delete message with unexisting id => 404
- [x] TEST: edit message with PROPPATCH oc:message
- [x] TEST: edit non-existing message with PROPPATCH => 404
- [ ] TEST: post message as share recipient :no_entry_sign: FAIL
- [ ] TEST: edit message as share recipient :no_entry_sign: FAIL

- [ ] BUG: "topMostParentId", "parent" is 0 but there is no comment with that id, should the property be empty instead ?
- [ ] BUG: cannot POST comment on shared file as recipient

@blizzz let's have a chat about whether we want 404 or empty response when no comments exist in some of the cases

blizzz commented:
> BUG: POST to a non-existing file id "http://root:admin@localhost/owncloud/remote.php/dav/comments/files/100 " must return 404, not 403 ?

The original thought why I was returning 403 was to avoid the user to be able to guess whether a file id that does not belong to the user exists or not, that was quite early in development and related to that state back then. Rethinking this, 404 makes more sense now.

blizzz committed:
smaller changes related to reviews

blizzz committed:
utilize elementMap for parsing DateTime string input argument for REPORT method

blizzz commented:
> BUG or misunderstanding: if file with id "9" has no comments attached, I'd expect that if I do a PROPFIND on http://root:admin@localhost/owncloud/remote.php/dav/comments/files/9 it would return 404, at least that's how I did with system tags.

My interpretation, if it would return 404, is that the file is not there instead of lacking comments.

PVince81 commented:
> My interpretation, if it would return 404, is that the file is not there instead of lacking comments.

Fine by me

blizzz committed:
return 404 instead of 403 when attempting to create a comment on an not existing or unavailable file

blizzz committed:
Revert "utilize elementMap for parsing DateTime string input argument for REPORT method"

This reverts commit 2027bf2686ee077fb2ad65f7a91f920c96b3fd85.

blizzz commented:
> BUG: "topMostParentId", "parent" is 0 but there is no comment with that id, should the property be empty instead ?

I do not see the use. Also, this will because only necessary if at some point we (or someone) needs to have full-thread support this becomes relevant. Not that it should be broken, but I don't see anything wrong. If the parent is 0 it's a thread head, so any client should and would not need to bother with this field.

blizzz committed:
adjust unit tests

blizzz committed:
files_sharing requires int as file id

blizzz commented:
@PVince81 I think I addressed your points, please recheck.

![](http://images.bigoven.com/image/upload/t_recipe-256/blueberry-pancakes-28bba1.jpg)
(and enjoy this digital pancake)

PVince81 commented:
Retested, all good now :+1: 

Needs a second reviewer @nickvergessen @icewind1991 @DeepDiver1975

icewind1991 commented:
Looks good as far as I can see :+1:

blizzz commented:
Thanks guys for support inbetween and reviews of course :bouquet:
